tests/tests_launch/test_launch_gcp.py

```diff
@@ -0,0 +1,222 @@
+        },
+    }
+    run = launch.run(**kwargs)
+    assert run.id == job_dict["name"]
+    assert run.name == job_dict["display_name"]
+    assert run.gcp_region == job_dict["location"]
+    assert run.gcp_project == job_dict["project"]
+    assert run.get_status().state == "finished"
+    assert run.wait()
+
+
+def test_vertex_options(test_settings, monkeypatch, mocked_fetchable_git_repo):
+    job_dict = setup_mock_aiplatform(monkeypatch)
+
+    api = wandb.sdk.internal.internal_api.Api(
+        default_settings=test_settings, load_settings=False
+    )
+    uri = "https://wandb.ai/mock_server_entity/test/runs/1"
+    kwargs = {
+        "uri": uri,
+        "api": api,
+        "resource": "gcp-vertex",
+        "entity": "mock_server_entity",
+        "project": "test",
+        "resource_args": {},
+    }
+    try:
+        launch.run(**kwargs)
+    except LaunchError as e:
+        assert "Vertex requires a staging bucket" in str(e)
+
+    kwargs["resource_args"]["gcp_staging_bucket"] = "test-bucket"
+    try:
+        launch.run(**kwargs)
+    except LaunchError as e:
+        assert "Vertex requires an Artifact Registry repository" in str(e)
+
+
+def test_vertex_supplied_docker_image(
+    test_settings, monkeypatch, mocked_fetchable_git_repo
+):
+    job_dict = setup_mock_aiplatform(monkeypatch)
+
+    def patched_pull_docker_image(docker_image):
+        return  # noop
+
+    def patched_docker_image_inspect(image):
+        return {
+            "ContainerConfig": {"WorkingDir": "/", "Env": [],},
+        }
+
+    monkeypatch.setattr(
+        "wandb.sdk.launch.runner.gcp_vertex.pull_docker_image",
+        lambda docker_image: patched_pull_docker_image(docker_image),
+    )
+    monkeypatch.setattr(
+        "wandb.sdk.launch.runner.gcp_vertex.docker_image_inspect",
+        lambda docker_image: patched_docker_image_inspect(docker_image),
+    )
+
+    api = wandb.sdk.internal.internal_api.Api(
+        default_settings=test_settings, load_settings=False
+    )
+    uri = "https://wandb.ai/mock_server_entity/test/runs/1"
+    kwargs = {
+        "uri": uri,
+        "api": api,
+        "resource": "gcp-vertex",
+        "entity": "mock_server_entity",
+        "project": "test",
+        "docker_image": "test:tag",
+        "resource_args": {
+            "gcp_staging_bucket": "test-bucket",
+            "gcp_artifact_repo": "test_repo",
+        },
+    }
+    run = launch.run(**kwargs)
+    assert run.id == job_dict["name"]
+    assert run.name == job_dict["display_name"]
+    assert run.gcp_region == job_dict["location"]
+    assert run.gcp_project == job_dict["project"]
+    assert run.get_status().state == "finished"
+
+
+def test_run_shell():
+    assert run_shell(["echo", "hello"]) == "hello"
+
+
+def test_get_gcp_config(monkeypatch):
+    def mock_gcp_config(args, stdout):
+        return mockdict(
+            {
+                "stdout": bytes(
+                    """
+is_active: true
```

Comment:
nit: indentation

Reply:
this is intentional — triple quotes preserves indentation so it needs to be left aligned. moved it out of this dict so it looks less ugly

Reply:
Oh, interesting didn't know that